File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        
        # store the frequency of each character
        frequencies = [0] * 26
        for c in s:
            frequencies[ord(c) - 97] += 1
        
        ans = 0
        # Use a set to store the frequencies we've already seen
        seen = set()
        
        for i in range(26):
            # Keep decrementing the frequencies we've already seen
            while frequencies[i] and frequencies[i] in seen:
                frequencies[i] -= 1
                ans += 1
                
            # Add the newly occupied frequency to the set
            seen.add(frequencies[i])
        
        return ans
        
                    
        

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.cpp
class Solution {
public:
    int minDeletions(string s) {
        
        // store the frequency orf each character
        vector<int> frequencies(26);
        for (char& c : s){
            ++frequencies[c - 'a'];
        }
        
        // Use a set to store the frequencies we've already seen
        unordered_set<int> seen;
        int ans = 0;
        for (int i = 0; i < 26; ++i){
            // keep decrementing the frequency until it is unique
            while (frequencies[i] && seen.find(frequencies[i]) != seen.end()){
                --frequencies[i];
                ++ans;
            }
            // add the newly occupied frequency to the set
            seen.insert(frequencies[i]);
        }
        
        return ans;
    }
};